server/src/main/java/org/elasticsearch/repositories/RepositoryData.java

```diff
@@ -980,6 +953,23 @@ private static void parseSnapshots(
         }
     }
 
+    private static IndexVersion parseIndexVersion(XContentParser.Token token, XContentParser parser) throws IOException {
+        switch (token) {
+            case VALUE_STRING -> {
+                Version v = Version.fromString(parser.text());
+                assert v.before(Version.V_8_10_0);
+                return IndexVersion.fromId(v.id);
+            }
+            case VALUE_NUMBER -> {
+                return IndexVersion.fromId(parser.intValue());
+            }
+            default -> {
+                XContentParserUtils.throwUnknownToken(parser.currentToken(), parser);
+                throw new AssertionError(); // never reached
+            }
+        }
+    }
```

Comment:
How about:

```java    
private static IndexVersion parseIndexVersion(XContentParser.Token token, XContentParser parser) throws IOException {
        if (token == XContentParser.Token.VALUE_NUMBER) {
            return IndexVersion.fromId(parser.intValue());
        } else {
            XContentParserUtils.ensureExpectedToken(XContentParser.Token.VALUE_STRING, token, parser);
            Version v = Version.fromString(parser.text());
            assert v.before(Version.V_8_10_0);
            return IndexVersion.fromId(v.id);
        }
    }
```

This should be faster (which does actually matter a little here) and avoids the confusing throwing of `AssertionError`?

Reply:
Hmm, or maybe change `ensureExpectedToken` to return an `Exception`, that can be used in a `throw` (even if the method itself throws directly), to keep the java compiler happy?

    switch {
        ...
        default -> throw XContentParserUtils.ensureExpectedToken(...);
    }

Reply:
Not sure that would be a good idea, it's definitely slower here I think and it's more code. The current form of that method I think is what compiles to the fastest code and is a convenient one-liner that can be used like an assertion.
Also that method is used 300+ times, we can't just change it :)

Reply:
OK, changed to suggestion